internal/internal_worker_base.go

```diff
@@ -332,61 +347,100 @@ func (bw *baseWorker) runPoller() {
 	defer bw.stopWG.Done()
 	bw.metricsHandler.Counter(metrics.PollerStartCounter).Inc(1)
 
+	ctx, cancelfn := context.WithCancel(context.Background())
+	reserveChan := make(chan *SlotPermit)
+
 	for {
+		bw.stopWG.Add(1)
+		go func() {
+			defer bw.stopWG.Done()
+			s, err := bw.slotSupplier.ReserveSlot(ctx, &bw.options.slotReservationData)
+			if err != nil {
+				if !errors.Is(err, context.Canceled) {
+					bw.logger.Error(fmt.Sprintf("Error while trying to reserve slot: %v", err))
+				} else {
+					close(reserveChan)
+				}
+				return
+			}
+			select {
+			case reserveChan <- s:
+			case <-ctx.Done():
+				bw.releaseSlot(s, "worker shutting down")
+			}
+		}()
+
 		select {
 		case <-bw.stopCh:
+			cancelfn()
 			return
-		case <-bw.pollerRequestCh:
+		case permit := <-reserveChan:
+			if permit == nil { // There was an error reserving a slot
+				// Avoid spamming reserve hard in the event it's constantly failing
+				time.Sleep(time.Second)
+				continue
```

Comment:
IIUC the channel can never be used again at this point correct, it is closed. Which means this select call will _always_ resolve immediately with a nil value. So why `continue`?

Reply:
@Sushisource I agree this looks like a bug

Reply:
This channel doesn't get closed any more